File: payments/src/main/java/com/volume/shared/infrastructure/rest/RestErrorResponse.java
package com.volume.shared.infrastructure.rest;

import com.volume.shared.domain.DomainException;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;
import lombok.Value;
import org.springframework.cloud.sleuth.Tracer;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

@Value
@NoArgsConstructor(force = true, access = AccessLevel.PRIVATE)
@AllArgsConstructor
public class RestErrorResponse {
    private final HttpStatus status;
    private final String message;
    private final String traceId;

    public static ResponseEntity<RestErrorResponse> fromException(HttpStatus status, DomainException exception, Tracer tracer) {
        return ResponseEntity.status(status).body(new RestErrorResponse(status, exception.getMessage(), tracer.currentSpan().context().traceId()));
    }
}

File: payments/src/main/java/com/volume/payments/shared/infrastructure/rest/RestErrorResponse.kt
package com.example.aggregatedemojava.utilities.infrastructure.rest

import com.volume.payments.shared.domain.DomainException
import org.springframework.cloud.sleuth.Tracer
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity

/**
 * TODO: I wonder if there is a way to make message property non-nullable
 */
data class RestErrorResponse(val status: HttpStatus, val message: String?, val traceId: String?) {
    companion object {
        fun fromException(status: HttpStatus, exception: DomainException, tracer: Tracer) : ResponseEntity<RestErrorResponse> {
            return ResponseEntity
                .status(status)
                .body(
                    RestErrorResponse(
                        status,
                        exception.message,
                        tracer.currentSpan()?.context()?.traceId())
                )
        }
    }
}

